File: gob/optimizers/cpp_optimizers/src/ECP.cc
/*
 * Created in 2025 by Gaëtan Serré
 */

#include "ECP.hh"

bool decision(
    const vector<dyn_vector> &points,
    const vector<double> &values,
    const dyn_vector &x,
    const double &epsilon,
    const function<double(dyn_vector x)> &f)
{
  double max_values = max_vec(values);
  vector<double> norms(points.size());
  for (int i = 0; i < points.size(); i++)
  {
    norms[i] = values[i] + epsilon * (x - points[i]).norm();
  }
  return max_values <= min_vec(norms);
}

result_eigen ECP::minimize(function<double(dyn_vector x)> f)
{
  vector<dyn_vector> points;
  vector<double> values;
  dyn_vector x = unif_random_vector(this->re, this->bounds);
  points.push_back(x);
  values.push_back(-f(x));
  int t = 1, h1 = 1, h2 = 0;
  while (t < this->n_eval)
  {
    x = unif_random_vector(this->re, this->bounds);
    h2++;
    if ((h2 - h1) > this->C)
    {
      this->epsilon = this->epsilon * this->theta;
      h2 = 0;
    }
    if (decision(points, values, x, this->epsilon, f))
    {
      points.push_back(x);
      values.push_back(-f(x));
      t++;
      h1 = h2;
      this->epsilon = this->epsilon * this->theta;
      h2 = 0;

      if (this->has_stop_criterion && -values.back() <= this->stop_criterion)
      {
        return {x, -values.back()};
      }
    }
  }

  int best_idx = argmax_vec(values);

  return {points[best_idx], -values[best_idx]};
}

File: gob/optimizers/ECP.py
#
# Created in 2024 by Gaëtan Serré
#

from .optimizer import Optimizer
from .cpp_optimizers import ECP as C_ECP


class ECP(Optimizer):
    def __init__(
        self,
        bounds,
        n_eval=50,
        epsilon=1e-2,
        theta_init=1.001,
        C=1000,
        max_trials=10_000_000,
        trust_region_radius=0.1,
        bobyqa_eval=10,
        verbose=False,
    ):
        super().__init__("ECP+TR", bounds)

        if n_eval < bobyqa_eval:
            bobyqa_eval = n_eval
            n_eval = 1
        else:
            n_eval = n_eval // bobyqa_eval

        self.c_opt = C_ECP(
            bounds,
            n_eval,
            epsilon,
            theta_init,
            C,
            max_trials,
            trust_region_radius,
            bobyqa_eval,
            verbose,
        )

    def minimize(self, f):
        return self.c_opt.minimize(f)

    def set_stop_criterion(self, stop_criterion):
        self.c_opt.set_stop_criterion(stop_criterion)

    def __del__(self):
        del self.c_opt
